crates/core/src/registry.rs

```diff
@@ -333,91 +347,160 @@ impl RegistryClient {
-            }
-        }
-
-        let response = request.send().await?;
-
-        if !response.status().is_success() {
-            let status = response.status();
-            let error_text = response.text().await.unwrap_or_default();
-            return Err(RegistryError::DownloadFailed {
-                status: status.into(),
-                message: error_text,
-            });
-        }
-
-        let package_data = response.bytes().await?;
-
-        // Check if the response is a JSON redirect with download_url
-        if package_data.len() >= 2 {
-            let magic = &package_data[0..2];
-            if magic != [0x1f, 0x8b] {
-                // Check if this is a JSON response with a download_url
-                if let Ok(text) = std::str::from_utf8(&package_data) {
-                    if text.trim().starts_with('{') {
-                        // Try to parse as JSON to get the actual download URL
-                        if let Ok(download_response) =
-                            serde_json::from_str::<DownloadResponse>(text)
-                        {
-                            debug!(
-                                "Server returned download URL: {}",
-                                download_response.download_url
-                            );
-
-                            // Download from the actual URL
-                            let actual_response = self
-                                .client
-                                .get(&download_response.download_url)
-                                .send()
-                                .await?;
-
-                            if !actual_response.status().is_success() {
-                                let status = actual_response.status();
-                                let error_text = actual_response.text().await.unwrap_or_default();
-                                return Err(RegistryError::CdnDownloadFailed {
-                                    status: status.into(),
-                                    message: error_text,
-                                });
-                            }
+                let access_token = auth.tokens.access_token;
+                let head_response = self
+                    .client
+                    .head(&download_url)
+                    .header("Authorization", format!("Bearer {access_token}"))
+                    .send()
+                    .await
+                    .map_err(|e| RegistryError::DownloadFailed {
+                        status: e
+                            .status()
+                            .unwrap_or(reqwest::StatusCode::INTERNAL_SERVER_ERROR)
+                            .into(),
+                        message: e.to_string(),
+                    })?;
+
+                let total_size = head_response
+                    .headers()
+                    .get(CONTENT_LENGTH)
+                    .and_then(|val| val.to_str().ok()?.parse().ok())
+                    .unwrap_or(0);
```

Comment:
should we default to a non-zero value like 1 so we don't run into a division by zero error?